File: app/src/main/java/com/pr0gramm/app/ui/dialogs/SearchUserDialog.java
package com.pr0gramm.app.ui.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.widget.AutoCompleteTextView;

import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.R;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.services.UserService;
import com.pr0gramm.app.services.UserSuggestionService;
import com.pr0gramm.app.ui.DialogBuilder;
import com.pr0gramm.app.ui.UsernameAutoCompleteAdapter;
import com.pr0gramm.app.ui.base.BaseDialogFragment;
import com.pr0gramm.app.ui.fragments.BusyDialogFragment;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;

import butterknife.BindView;

/**
 */
public class SearchUserDialog extends BaseDialogFragment {
    private static final Logger logger = LoggerFactory.getLogger("SearchUserDialog");

    @BindView(R.id.username)
    AutoCompleteTextView inputView;

    @Inject
    UserService userService;

    @Inject
    UserSuggestionService suggestionService;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return DialogBuilder.start(getContext())
                .layout(R.layout.search_user_dialog)
                .positive(R.string.action_search_simple, di -> onSearchClicked())
                .negative(Dialog::dismiss)
                .noAutoDismiss()
                .build();
    }

    @Override
    protected void onDialogViewCreated() {
        inputView.setAdapter(new UsernameAutoCompleteAdapter(suggestionService, getThemedContext(),
                "", android.R.layout.simple_dropdown_item_1line));

    }

    private void onSearchClicked() {
        String username = inputView.getText().toString().trim();

        userService.info(username)
                .compose(bindToLifecycleAsync())
                .lift(BusyDialogFragment.busyDialog(this))
                .subscribe(this::onSearchSuccess, this::onSearchFailure);
    }

    private void onSearchSuccess(Api.Info info) {
        logger.info("Found user info: {} {}", info.getUser().getId(), info.getUser().getName());

        Fragment parent = getParentFragment();
        if (parent instanceof Listener) {
            ((Listener) parent).onUserInfo(info);
        }

        dismissAllowingStateLoss();
    }

    private void onSearchFailure(Throwable throwable) {
        inputView.setError(getString(R.string.user_not_found));
    }

    @Override
    protected void injectComponent(ActivityComponent activityComponent) {
        activityComponent.inject(this);
    }

    public interface Listener {
        void onUserInfo(Api.Info info);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/dialogs/SearchUserDialog.kt
package com.pr0gramm.app.ui.dialogs

import android.app.Dialog
import android.os.Bundle
import android.widget.AutoCompleteTextView
import com.pr0gramm.app.R
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.services.UserService
import com.pr0gramm.app.services.UserSuggestionService
import com.pr0gramm.app.ui.UsernameAutoCompleteAdapter
import com.pr0gramm.app.ui.base.BaseDialogFragment
import com.pr0gramm.app.ui.base.bindView
import com.pr0gramm.app.ui.base.launchWhenStarted
import com.pr0gramm.app.ui.dialog
import com.pr0gramm.app.util.di.instance
import kotlinx.coroutines.CancellationException

/**
 */
class SearchUserDialog : BaseDialogFragment("SearchUserDialog") {
    private val userService: UserService by instance()
    private val suggestionService: UserSuggestionService by instance()

    private val inputView: AutoCompleteTextView by bindView(R.id.username)

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return dialog(requireContext()) {
            layout(R.layout.search_user_dialog)
            positive(R.string.action_search_simple) { onSearchClicked() }
            negative { dismiss() }
            noAutoDismiss()
        }
    }

    override fun onDialogViewCreated() {
        inputView.setAdapter(UsernameAutoCompleteAdapter(suggestionService, themedContext,
                android.R.layout.simple_dropdown_item_1line, ""))

    }

    private fun onSearchClicked() {
        val username = inputView.text.toString().trim()

        launchWhenStarted(busyIndicator = true) {
            try {
                onSearchSuccess(userService.info(username))
            } catch (err: CancellationException) {
            } catch (err: Exception) {
                onSearchFailure()
            }
        }
    }

    private fun onSearchSuccess(info: Api.Info) {
        logger.info { "Found user info: ${info.user.id} ${info.user.name}" }

        (parentFragment as? Listener)?.onUserInfo(info)
        dismissAllowingStateLoss()
    }

    private fun onSearchFailure() {
        inputView.error = getString(R.string.user_not_found)
    }

    interface Listener {
        fun onUserInfo(info: Api.Info)
    }
}
